### ISSUE TITLE
Serious bug regarding custom choice field and its value in cdata table

### ISSUE THREAD

**Issue by esamo** [osTicket/osTicket#4071]
I had a problem that every new ticket had wrong values in cdata table on custom choice fields.
Turns out instead of value the key was stored, but the key was always '0' not depending on the actual choice.

Here's exact place in the code
http://take.ms/pfW2e

But the fix should be different and tested with multioption fields etc.
http://take.ms/GudHoj

---

Config & versions
http://take.ms/xmqOJ

**Comment by esamo** [comment]
Maybe there's place that calls getKeys instead of toString

**Comment by mfelber** [comment]
I’m not good in this things but I can recall that somebody has mentioned a behavior with writing zeros into the database instead of correct values. Maybe there is a similar reason for this case too: https://github.com/osTicket/osTicket/issues/3620

**Comment by esamo** [comment]
Yeah actually getKeys is working correctly, but somewhere it is called instead of proper function like toString.

When cdata is restored, it shows 'string_value,string_value' (with a comma) in the custom choice field, but when it is saved during new ticket creation, it stores just '0', which says something is totally mishandled in custom choice field save into cdata table.

I lost too much time already to go figuring out that :P.
So my 'fix' (not really a fix) is that getKeys returns 'string_value,string_value'.

**Comment by JediKev** [comment]
@esamo Okay, so it took me a while to figure out what was happening. Could you post a screenshot of your Choice Field's choices? _**Admin Panel > Manage > Forms > Ticket Details Form > Choices Field > Config**_

**Comment by esamo** [comment]
Here is custom choice config
http://take.ms/DftQf

Here is how it looks in DB
http://take.ms/OLNNb

I just noticed the note above which instructs to configure the list with key:value params not value alone.  Ooops :D.
But shouldn't be error displayed if the system won't function as expected? It saves the config fine.

**Comment by JediKev** [comment]
Yea that's your issue, you need to have keys on those. I have a [pull request that "fixes" this issue](https://github.com/osTicket/osTicket/pull/4007). You're supposed to have keys for all values in order for searching and exporting to work for Choice Fields. So the pull "forces" keys for the choices so everything will work as expected. Once you have keys and you update a choice on a ticket or create a new ticket with the choice value you'll be able to see the correct 'key' stored in the cdata table. This way for searching and exporting it'll pull the correct value by the key. 👍 

**Note:** If you have more than two choices then it's better to have the field as a list instead (it just works better).

Cheers.

**Comment by Lurch440** [comment]
Regression in 1.11
I get the "Each choice requires a key ..." in red.

My list in the form of:
1:value
2:yada yada ..

**Comment by Lurch440** [comment]
Seems that "(" is considered an illegal character.  If the list contains even one, the submission fails.

**Comment by JediKev** [comment]
@Lurch440 

You need this:
https://github.com/osTicket/osTicket/pull/4847

Cheers.
